### ISSUE TITLE
torch.ops._caffe2 linked no functions in windows -->RuntimeError [enforce fail at roi_align_op.h:39] Not Implemented

### ISSUE THREAD

**Issue by SpaceView** [pytorch/pytorch#41896]
## 🐛 Bug
_caffe2 has linked no functions, and I think this is a serious problem in pytorch, thus I openned this new issue report. Hope anybody familiar with pytorch compilation in windows could help...

I refered to https://github.com/pytorch/pytorch/issues/35547, but I don't know how this problem is solved, since I'm facing the exact errors when deploying detectron2 (use the detectron2 caffe2_converter.py file, everything follows the dectectron2 deployment instruction).
I tried torch 1.5, 1.5.1, and in addition I cloned the repo 2 days ago to build the pytorch successfully on win10, detectron setup is also OK, and I tried to use vscode to see the debug, as below,
Traceback (most recent call last):
  File "D:\Anaconda3\envs\tchdev\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "D:\Anaconda3\envs\tchdev\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "c:\Users\Administrator\.vscode\extensions\ms-python.python-2020.7.94776\pythonFiles\lib\python\debugpy\__main__.py", line 45, in <module>
    cli.main()
  File "c:\Users\Administrator\.vscode\extensions\ms-python.python-2020.7.94776\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 430, in main
    run()
  File "c:\Users\Administrator\.vscode\extensions\ms-python.python-2020.7.94776\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 267, in run_file
    runpy.run_path(options.target, run_name=compat.force_str("__main__"))
  File "D:\Anaconda3\envs\tchdev\lib\runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "D:\Anaconda3\envs\tchdev\lib\runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "D:\Anaconda3\envs\tchdev\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "d:\devPytorch\detectron2_deploy\caffe2_converter.py", line 65, in <module>
    caffe2_model = tracer.export_caffe2()
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\detectron2-0.2-py3.7-win-amd64.egg\detectron2\export\api.py", line 112, in export_caffe2
    predict_net, init_net = export_caffe2_detection_model(model, inputs)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\detectron2-0.2-py3.7-win-amd64.egg\detectron2\export\caffe2_export.py", line 143, in export_caffe2_detection_model
    onnx_model = export_onnx_model(model, (tensor_inputs,))
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\detectron2-0.2-py3.7-win-amd64.egg\detectron2\export\caffe2_export.py", line 60, in export_onnx_model
    operator_export_type=OperatorExportTypes.ONNX_ATEN_FALLBACK,
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\onnx\__init__.py", line 168, in export
    custom_opsets, enable_onnx_checker, use_external_data_format)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\onnx\utils.py", line 69, in export
    use_external_data_format=use_external_data_format)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\onnx\utils.py", line 488, in _export
    fixed_batch_size=fixed_batch_size)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\onnx\utils.py", line 334, in _model_to_graph
    graph, torch_out = _trace_and_get_graph_from_model(model, args, training)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\onnx\utils.py", line 291, in _trace_and_get_graph_from_model
    torch.jit._get_trace_graph(model, args, _force_outplace=False, _return_inputs_states=True)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\jit\__init__.py", line 278, in _get_trace_graph
    outs = ONNXTracedModule(f, _force_outplace, return_inputs, _return_inputs_states)(*args, **kwargs)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\nn\modules\module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\jit\__init__.py", line 361, in forward
    self._force_outplace,
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\jit\__init__.py", line 348, in wrapper
    outs.append(self.inner(*trace_inputs))
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\nn\modules\module.py", line 548, in __call__
    result = self._slow_forward(*input, **kwargs)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\nn\modules\module.py", line 534, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "D:\Anaconda3\envs\tchdev\lib\contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\detectron2-0.2-py3.7-win-amd64.egg\detectron2\export\caffe2_modeling.py", line 272, in forward
    detector_results, _ = self._wrapped_model.roi_heads(images, features, proposals)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\nn\modules\module.py", line 548, in __call__
    result = self._slow_forward(*input, **kwargs)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\nn\modules\module.py", line 534, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\detectron2-0.2-py3.7-win-amd64.egg\detectron2\modeling\roi_heads\roi_heads.py", line 674, in forward
    pred_instances = self._forward_box(features, proposals)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\detectron2-0.2-py3.7-win-amd64.egg\detectron2\modeling\roi_heads\roi_heads.py", line 727, in _forward_box
    box_features = self.box_pooler(features, [x.proposal_boxes for x in proposals])
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\nn\modules\module.py", line 548, in __call__
    result = self._slow_forward(*input, **kwargs)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\torch\nn\modules\module.py", line 534, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "D:\Anaconda3\envs\tchdev\lib\site-packages\detectron2-0.2-py3.7-win-amd64.egg\detectron2\export\c10.py", line 336, in forward
    aligned=aligned,
RuntimeError: [enforce fail at roi_align_op.h:39] . Not Implemented.
<-------------------------------------------------------------------->
and I tried to see what is going on in python, as below
(tchdev) D:\devPytorch>python
Python 3.7.6 | packaged by conda-forge | (default, Jun  1 2020, 18:11:50) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch as tch
>>> dir(tch.ops._caffe2)
['__doc__', '__loader__', '__name__', '__package__', '__spec__', 'name']
>>> dir(tch.ops._caffe2.__name__)
['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']
>>>
<-------------------------------------------------------------------->
it seems indeed roi_align_op is not linked in, and _caffe2 doesn't see to have any useful functions!
but why, and how can we compile the _caffe2 functions into the lib/dll ?

## To Reproduce

1.  use any version of pytorch in windows 10
1.  install detectron2 as given by  https://github.com/facebookresearch/detectron2
1.  python caffe2_converter.py --config-file ../../configs/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml --output ./caffe2_model --run-eval MODEL.WEIGHTS detectron2://COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x/137849600/model_final_f10217.pkl MODEL.DEVICE cpu

## Expected behavior

no error

## Environment

 - PyTorch Version (1.5, 1.5.1 and the one I compiled in July 22):
 - OS (win 10):
 - How you installed PyTorch (1.5, 1.5.1 with conda, the compiled one with python setup.py install):
 - Build command you used (python setup.py install):
 - Python version:
 - CUDA/cuDNN version 10.1:
 - GPU models and configuration:
 - Any other relevant information:

## Additional context

cc @malfet @seemethere @walterddr @peterjc123 @maxluk @nbcsm @guyang3532 @gunandrose4u @smartcat2010 @mszhanyi

**Comment by SpaceView** [comment]
Sorry if I didn't describe it correctly, not that familiar with pytorch details yet.
Addition information for my above Bug. I checked my ubuntu18.04 yesterday, e.g.
>>>dir(torch.ops._caffe2)
['AliasWithName', 'CollectRpnProposals', 'DistributeFpnProposals', 'GenerateProposals', 'ResizeNearest', '__doc__', '__loader__', '__name__', '__package__', '__spec__', 'name']

then after executing the 325th line in c10.py --> "c2_roi_align = torch.ops._caffe2.RoIAlign", I got the following
>>>c2_roi_align
<built-in method RoIAlign of PyCapsule object at 0x7fb11a71ef00>
>>>dir(torch.ops._caffe2)
['AliasWithName', 'CollectRpnProposals', 'DistributeFpnProposals', 'GenerateProposals', 'ResizeNearest', 'RoIAlign', '__doc__', '__loader__', '__name__', '__package__', '__spec__', 'name']
You see, RoIAlign is here now!

It seems that these are dynamically linked pycapsule objects, it works in Linux but NOT under windows. But why? and how could we fix it?
Greatly appreciate someone can give me some clues.

**Comment by peterjc123** [comment]
@SpaceView Could you please provide the result with the nightly packages? I believe it should not be `RuntimeError: [enforce fail at roi_align_op.h:39] . Not Implemented.` anymore. This is a known issue in PyTorch 1.5.1.

**Comment by freedenS** [comment]
i test libtorch1.6 from binary, but it still be RuntimeError: [enforce fail at roi_align_op.h:39] . Not Implemented.

**Comment by freedenS** [comment]
Here is my code and pt file. https://drive.google.com/file/d/18t48TcG_kd53ILDiy3PQ-KzyFeK7im_G/view?usp=sharing
```
#include <iostream>
#include <string>
#include <torch/script.h>

using namespace std;

int main() {
  auto module = torch::jit::load("./pt/alignr.pt");

  auto input = torch::arange(25, torch::kFloat32).reshape({5,5}).unsqueeze(0).unsqueeze(0);
  auto rois = torch::ones({ 1,6 }, torch::kFloat32);
  rois[0][0] = 0;
  rois[0][1] = 2;
  rois[0][2] = 2;
  rois[0][3] = 2;
  rois[0][4] = 2;
  rois[0][5] = 90;

  std::vector<torch::jit::IValue> inputs;
  inputs.push_back(input);
  inputs.push_back(rois);

  try
  {
	  auto output = module.forward(inputs).toTensor();
  }
  catch (exception& e)
  {
	  cout << e.what() << endl;
  }

  return 0;
}
```

**Comment by mszhanyi** [comment]
The root cause is that the specification template function couldn't be exported in dll because it's in the cpp file. I tested that it could be exported only if it's defined in the header file. I raised [an issue](https://developercommunity.visualstudio.com/content/problem/1193545/template-class-specification-member-function-could.html) to VC team.  And I wrote a small demo to reproduce it.  @bwignall @SuperIRabbit @linbinyu @EscapeZero @orionr I  found you were the author of the code and I think other ops would have the similar issue in windows. One workaround is to move the specification template function to .h but it's not a small cost and break current code style.
